symbolic-base/src/ZkFold/Protocol/IVC/Oracle.hs

```diff
@@ -1,75 +1,92 @@
-{-# LANGUAGE AllowAmbiguousTypes  #-}
-{-# LANGUAGE TypeOperators        #-}
-{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DerivingVia         #-}
+{-# LANGUAGE TypeOperators       #-}
 
 module ZkFold.Protocol.IVC.Oracle where
 
-import qualified Data.Vector                                   as V
+import           Data.Foldable                                 (Foldable, foldMap)
+import           Data.Function                                 ((.))
+import           Data.List                                     ((++))
 import           GHC.Generics
-import           Prelude                                       (map, (.))
-import qualified Prelude                                       as P
 
-import           ZkFold.Algebra.Class
+import           ZkFold.Algebra.Class                          (Ring, zero)
+import           ZkFold.Data.Vector                            (Vector)
 import           ZkFold.Symbolic.Algorithm.Hash.MiMC           (mimcHashN')
 import           ZkFold.Symbolic.Algorithm.Hash.MiMC.Constants (mimcConstants)
 import           ZkFold.Symbolic.Class                         (Arithmetic)
 
+----------------------------- OracleSource class -------------------------------
+
 -- TODO: add more specific instances for efficiency
 
-class HashAlgorithm algo a where
-    hash :: [a] -> a
+-- | @OracleSource a b@ links together base field @a@ and source of randomness @b@ to use in 'oracle'.
+--
+-- 'Generic' deriving via @b@ is available.
+class OracleSource a b where
+    source :: b -> [a]
+    -- ^ Extracts random seed from the source.
+    default source :: (Generic b, GOracleSource a (Rep b)) => b -> [a]
+    source = gsource . from
+
+instance OracleSource a a where
+    source = (:[])
+
+instance (OracleSource a b, OracleSource a c) =>
+    OracleSource a (b, c)
+
+instance (OracleSource a b, OracleSource a c, OracleSource a d) =>
+    OracleSource a (b, c, d)
+
+instance ( OracleSource a b, OracleSource a c
+         , OracleSource a d, OracleSource a e) => OracleSource a (b, c, d, e)
 
-data MiMCHash a
-instance forall a. (Ring a, Arithmetic a) => HashAlgorithm (MiMCHash a) a where
-    hash :: forall x. (Arithmetic x, FromConstant a x) => [x] -> x
-    hash = mimcHashN' @a mimcConstants zero
+instance ( OracleSource a b, OracleSource a c, OracleSource a d
+         , OracleSource a e, OracleSource a f) => OracleSource a (b, c, d, e, f)
 
-class RandomOracle algo x a where
-    oracle :: x -> a
-    default oracle :: (Generic x, RandomOracle' algo (Rep x) a) => x -> a
-    oracle = oracle' @algo . from
+-- | A newtype to derive 'OracleSource' for any 'Foldable'
+-- as long as its element type is also 'OracleSource'.
+newtype FoldableSource f a = FoldableSource { foldableSource :: f a }
```

Comment:
I had to introduce this newtype to avoid overlapping (and **incoherent**, even) instances: `(a, b)` is `Foldable` in `b` as well, keep that in mind!

So actually this newtype is what we really need, because by using `FoldableSource f a` we signal that we drop any extra info in `f` which is not present via elements of `a`.